### DISCUSSION TITLE
Windows cmake build: mismatch detected for '_ITERATOR_DEBUG_LEVEL'

### DISCUSSION THREAD

**Discussion by brianheilig** [protobuf-c/protobuf-c#575] [Q&A] [answered]
While building protobuf-c using the provided cmake file I encountered the following linker error (the first of many similar errors):

```
libprotoc.lib(code_generator.cc.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't
match value '2' in c_bytes_field.obj [C:\Users\brian\Documents\protobuf-c\build-cmake\build\protoc-gen-c.vcxproj]
libprotoc.lib(code_generator.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease'
doesn't match value 'MTd_StaticDebug' in c_bytes_field.obj [C:\Users\brian\Documents\protobuf-c\build-cmake\build\proto
c-gen-c.vcxproj]
```

Here are the steps I followed and the build output. I cloned protobuf-c from master.

```
C:\Users\brian\Documents\protobuf-c\build-cmake> mkdir build && cd build
C:\Users\brian\Documents\protobuf-c\build-cmake\build> cmake ..
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.31.31104.0
-- The CXX compiler identification is MSVC 19.31.31104.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.31.31103/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.31.31103/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
C:/Users/brian/Documents/protobuf-c
-- Found Protobuf: C:/Users/brian/Documents/protobuf/lib/libprotobuf.lib (found version "3.21.12")
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/brian/Documents/protobuf-c/build-cmake/build

C:\Users\brian\Documents\protobuf-c\build-cmake\build> cmake --build .
```

It's a pretty basic usage of cmake, so I'm assuming it's a bug. Compilation works fine but linking doesn't. Here's what else I tried:

`cmake -DCMAKE_BUILD_TYPE=Release ..`
`cmake -G "Visual Studio 15 2017 Win64" ..` to select the Visual Studio 2017 tools.

**Comment by danyeaw** [comment, marked as answer]
Hi @brianheilig, did you compile Protobuf with `-DCMAKE_BUILD_TYPE=Release` as well? It looks like you are trying to do a static build and the build types need to match.

**Comment by brianheilig** [comment]
Yes, you are right. That was the problem. Thank you!

**Comment by brianheilig** [comment]
For anyone else arriving here, here are the exact steps I took. I followed the instructions at https://github.com/protocolbuffers/protobuf/blob/main/cmake/README.md, but here's exactly what I did. I'm sure there's a better way but I couldn't figure it out.

**Build Protobuf**
Download protobuf-all-21.12.zip from https://github.com/protocolbuffers/protobuf/releases/latest. Extract to \protobuf-21.12
Open a Visual Studio 2022 x64 Native Tools command prompt and CD to \protobuf-21.12.
mkdir build && mkdir install && cd build
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=\protobuf-21.12\install ..
cmake --build . --config Release
cmake --build . --config Release --target Install

**Build Protobuf-c**
Clone the protobuf-c git repository to directory \protobuf-c
@danyeaw I bumped the CMAKE_MINIMUM_REQUIRED version to 3.12 in \protobuf-c\build-cmake\CMakeLists.txt. This allows the Protobuf_ROOT to be specified. **See below**.
cd \protobuf-c\build-cmake
mkdir build && cd build
cmake -DCMAKE_BUILD_TYPE=Release **-DProtobuf_ROOT**=\UserData\z003hu6t\Documents\protobuf-21.12\install ..
cmake --build . --config Release
